coreax/inverses.py

```diff
@@ -0,0 +1,312 @@
+# © Crown Copyright GCHQ
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Classes and associated functionality to approximate regularised inverses of matrices.
+
+Given a matrix :math:`A\in\mathbb{R}^{n\times n}` and some regularisation parameter
+:math:`\lambda\in\mathbb{R}_{\ge 0}`, the regularised inverse of :math:`A` is
+:math:`(A + \lambda I_n)^{-1}` where :math:`I_n` is the :math:`n\times n` identity
+matrix.
+
+When a dataset is very large, methods which have to invert regularised kernel matrices
+evaluated on all of the data can become prohibitively expensive. To reduce this
+computational cost, such methods can instead be approximated (providing suitable
+approximation error can be achieved).
+
+The :class:`RegularisedInverseApproximator` in this module provides the functionality
+required to approximate the inverse of regularised matrices. Furthermore, this class
+allows for "block-inversion" where given an invertible matrix :math:`B`, the block
+array
+
+.. math::
+    A = \begin{bmatrix}
+        B & 0 & \dots & 0 \\
+        0 & 0 & \dots & 0 \\
+        \vdots & \ddots & \dots & \vdots \\
+        0 & 0 & \dots & 0
+    \end{bmatrix}
+
+where only the top-left block contains non-zero elements can be "inverted" to give
+
+.. math::
+    A^{-1} := \begin{bmatrix}
+        B^{-1} & 0 & \dots & 0 \\
+        0 & 0 & \dots & 0 \\
+        \vdots & \ddots & \dots & \vdots \\
+        0 & 0 & \dots & 0
+    \end{bmatrix}.
+
+This functionality allows iterative coreset algorithms which require inverting growing
+arrays to have fully static array shapes and thus be JIT-compilable in JAX.
```

Comment:
I meant it to be `compilable`, like is possible to compile.

Reply:
Ok, I see. Because `jax.jit` transforms a function (it acts on a python function) I tend to prefer the phrase 'JIT compatible' (compatible can be freely interchanged with transformable). That said, the main reason I would prefer not to use 'compilable' here is because it has no entry in the OED and can be easily misread as compliable (which I did).

Reply:
Makes sense, changed